source/tscript.js

```diff
@@ -2112,6 +2112,106 @@ let lib_canvas = {
 	};
 
 
+
+	let lib_audio = {
+		"source":`
+			namespace audio{
+				class MonoAudio
+				{
+					public:
+					constructor(buffer, sampleRate){}
+					function play(){}
+					function pause(){}
+					function setPlaybackRate(speed){}
+				}
+				class StereoAudio
+				{
+					public:
+					constructor(leftBuffer, rightBuffer, sampleRate){}
+					function play(){}
+					function pause(){}
+					function setPlaybackRate(speed){}
+				}
+			}
+		`,
+		"impl": {
+			"audio":{
+				"MonoAudio":{
+					"constructor":function(object, buffer, sampleRate){
+						if (! module.isDerivedFrom(buffer.type, module.typeid_array)) this.error("/argument-mismatch/am-1", ["buffer", "audio.MonoAudio.constructor", "array", module.displayname(buffer)]);
+						if(! module.isDerivedFrom(sampleRate.type, module.typeid_integer)) this.error("/argument-mismatch/am-1", ["sampleRate", "audio.MonoAudio.constructor", "integer", module.displayname(sampleRate)])	
+						let buf = this.service.audioContext.createBuffer(1, buffer.value.b.length, sampleRate.value.b);
+						let ar = buf.getChannelData(0);
+						//copy from buffer to js objects buffer
+						for(let i=0; i < ar.length; i++){
+							//maybe check if number is a real and between -1 and 1?
```

Comment:
Yes, please! I actually don't know what happens if we give JS a junk value at this point (I assume that the audio buffer is a typed array), so the user should get an error message if values are non-numeric or NaN. On the other hand, for the values I think that clipping to [-1, 1] would be more reasonable than raising an error.

Reply:
The array is typed and JS will convert anything invalid to NaN.
![image](https://user-images.githubusercontent.com/28152920/103919714-a71db280-5110-11eb-8c48-ae0213989a52.png)